docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,455 @@
+	// standard label selector semantics; if present but empty, it selects all namespaces.
+	//
+	// If podSelector is also set, then the NetworkQoS as a whole selects
+	// the pods matching podSelector in the namespaces selected by namespaceSelector.
+	// Otherwise it selects all pods in the namespaces selected by namespaceSelector.
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty" protobuf:"bytes,2,opt,name=namespaceSelector"`
+
+	// ipBlock defines policy on a particular IPBlock. If this field is set then
+	// neither of the other fields can be.
+	// +optional
+	IPBlock *networkingv1.IPBlock `json:"ipBlock,omitempty" protobuf:"bytes,3,rep,name=ipBlock"`
+}
+
+// Status defines the observed state of NetworkQoS
+type Status struct {
+	// A concise indication of whether the NetworkQoS resource is applied with success.
+	// +optional
+	Status string `json:"status,omitempty"`
+
+	// An array of condition objects indicating details about status of NetworkQoS object.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=networkqoses
+// +kubebuilder::singular=networkqos
+// NetworkQoSList contains a list of NetworkQoS
+type NetworkQoSList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []NetworkQoS `json:"items"`
+}
+```
+
+* Ensure backward compatibility support available for `EgressQoS` API.
+
+### Implementation Details
+
+The new controller is introduced in OVN-Kubernetes which would watch `NetworkQoS` in addition to `EgressQoS`, `Pod` and `Node` objects, which will create the relevant NetworkQoS objects and attach them to all of the node local switches in the cluster in OVN - resulting in the necessary flows to be programmed in OVS.
+
+In order to not create an OVN NetworkQoS object per pod in the namespace, the controller will also manage AddressSets.
+For each QoS rule specified in a given `NetworkQoS` it'll create an AddressSet, adding only the pods whose label matches the PodSelector to it, making sure that new/updated/deleted matching pods are also added/updated/deleted accordingly. Rules that do not have a PodSelector will leverage the namespace's AddressSet.
+
+Similarly when `NetworkQoS` is created for Pods secondary network, OVNK must create a new AddressSet for every QoS rule. When no pod selector is specified, then it must contain all of the pod's IP addresses that belong to the namespace and selected network. If only a set of pods are chosen via podSelector, then it must have IP addresses only for chosen pod(s).
+
+For example, using LGW mode and assuming there's a single node `node1` and the following `NetworkQoS` is created:
```

Comment:
how would things be different if we are in Shared Gateway Mode? 

i don't see any difference.. can we remove 'using LGW mode'?